src/serde.rs

```diff
@@ -0,0 +1,87 @@
+use serde::{ser::Serialize, Deserialize};
+
+use crate::{NodeId, NodeRef, Tree};
+
+#[derive(Debug)]
+struct SerNode<'a, T> {
+    value: &'a T,
+    children: Vec<SerNode<'a, T>>,
+}
+
+impl<'a, T> From<NodeRef<'a, T>> for SerNode<'a, T> {
+    fn from(node: NodeRef<'a, T>) -> Self {
+        let value: &T = node.value();
+        let children: Vec<SerNode<'a, T>> = node.children().map(SerNode::<'a, T>::from).collect();
+        SerNode { value, children }
+    }
+}
+
+impl<'a, T: Serialize> Serialize for SerNode<'a, T> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        (&self.value, &self.children).serialize(serializer)
+    }
+}
+
+impl<T: Serialize> Serialize for Tree<T> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        SerNode::from(self.root()).serialize(serializer)
+    }
+}
+
+#[derive(Debug)]
+struct DeserNode<T> {
+    value: T,
+    children: Vec<DeserNode<T>>,
+}
+
+impl<T> DeserNode<T> {
+    fn into_tree_node(self, tree: &mut Tree<T>, parent: NodeId) -> NodeId {
```

Comment:
Could we pass a `parent: NodeMut` here instead of tree and parent ID? This might avoid the repeated look-ups by ID.

Reply:
I think we could do that, let me try.

Reply:
Fixed